service/service.go

```diff
@@ -69,12 +74,72 @@ func New(config config.Config, serviceLogger *logging.ServiceLogger) (ProxyServi
 		Handler: mux,
 	}
 
+	// create database client
+	db, err := createDatabase(ctx, config, serviceLogger)
+
+	if err != nil {
+		return ProxyService{}, err
+	}
+
 	return ProxyService{
 		httpProxy:     server,
 		ServiceLogger: serviceLogger,
+		database:      db,
 	}, nil
 }
 
+func createDatabase(ctx context.Context, config config.Config, logger *logging.ServiceLogger) (*database.PostgresClient, error) {
+	databaseConfig := database.PostgresDatabaseConfig{
+		DatabaseName:        config.DatabaseName,
+		DatabaseEndpointURL: config.DatabaseEndpointURL,
+		DatabaseUserName:    config.DatabaseUserName,
+		DatabasePassword:    config.DatabasePassword,
+		SSLEnabled:          config.DatabaseSSLEnabled,
+		QueryLoggingEnabled: config.DatabaseQueryLoggingEnabled,
+		Logger:              logger,
+	}
+
+	serviceDatabase, err := database.NewPostgresClient(databaseConfig)
+
+	if err != nil {
+		logger.Error().Msg(fmt.Sprintf("error %s creating database using config %+v", err, databaseConfig))
+
+		return &database.PostgresClient{}, err
+	}
+
+	// TODO: move into `Run` method
+	// wait for database to be reachable
+	var databaseOnline bool
+	for !databaseOnline {
+		err = serviceDatabase.HealthCheck()
+
+		if err != nil {
+			logger.Debug().Msg("unable to connect to database, will retry in 1 second")
+
+			time.Sleep(1 * time.Second)
+
+			continue
+		}
+
+		logger.Debug().Msg("connected to database")
+
+		databaseOnline = true
+	}
+
+	logger.Debug().Msg("running migrations on database")
+
+	migrations, err := database.Migrate(ctx, serviceDatabase.DBConnection, *migrations.Migrations)
```

Comment:
Generally, it's best to not run migrations in the main command and have it be running able by a separate cmd and run as part of a deployment script.  Running it with the server makes certain changes very difficult and when more then one replica is deployed, there can be race conditions (I'm not familar with what locks upscale orm uses).

Reply:
locks/migration protection is one of the nice value adds of the orm, they have a lock table and a migration table, the migration process attempts to grab a lock before running any migration to prevent race conditions. I'll add a ticket for breaking the running of migrations out into a separate command.

Reply:
i was briefly trying to look into this. are you certain the orm handles the migration locking?

[the code](https://github.com/uptrace/bun/blob/master/migrate/migrator.go#L385-L406) and [this comment from the maintainer](https://github.com/uptrace/bun/issues/758#issuecomment-1386741076) make me think they expect us to call Lock / Unlock ourselves to manage the race condition.

`migrator.Lock` and `defer migrator.Unlock` are not called from anywhere in the `migrator.Migrate` method: https://github.com/uptrace/bun/blob/master/migrate/migrator.go#L132

Reply:
judging by the [migration example](https://github.com/uptrace/bun/blob/master/example/migrate/main.go#L61-L64), it looks like we can safely keep it here if we first manually secure a lock on the migrations table

Reply:
thanks @nddeluca @pirtleshell going to make running migrations keyed off an environment variable to give us control when they run and add locking (the older version of this library I used extensively - go-pg, did handle the migration lock and I wrongly assumed the newer version did too 😅 )